src/lib/operators/window_function_evaluator.cpp

```diff
@@ -0,0 +1,607 @@
+          case WindowFunction::Avg:
+            return _templated_on_execute<InputColumnType, WindowFunction::Avg>();
+          case WindowFunction::Count:
+            return _templated_on_execute<InputColumnType, WindowFunction::Count>();
+          case WindowFunction::Min:
+            return _templated_on_execute<InputColumnType, WindowFunction::Min>();
+          case WindowFunction::Max:
+            return _templated_on_execute<InputColumnType, WindowFunction::Max>();
+          default:
+            Fail("Unsupported window function.");
+        }
+      }();
+    } else {
+      Fail("Unsupported input column type for window function.");
+    }
+  });
+  return result;
+}
+
+namespace {
+
+template <typename T>
+T clamped_add(T lhs, T rhs, T max) {
+  return std::min(lhs + rhs, max);
+}
+
+template <typename T>
+T clamped_sub(T lhs, T rhs, T min) {
+  using U = std::make_signed_t<T>;
+  return static_cast<T>(std::max(static_cast<U>(static_cast<U>(lhs) - static_cast<U>(rhs)), static_cast<U>(min)));
+}
+
+}  // namespace
+
+template <typename InputColumnType, WindowFunction window_function>
+WindowFunctionEvaluator::ComputationStrategy WindowFunctionEvaluator::choose_computation_strategy() const {
+  const auto& frame = frame_description();
+  const auto is_prefix_frame = frame.start.unbounded && frame.start.type == FrameBoundType::Preceding &&
+                               !frame.end.unbounded && frame.end.type == FrameBoundType::CurrentRow;
+  Assert(is_prefix_frame || !is_rank_like(window_function), "Invalid frame for rank-like window function.");
+
+  if (is_prefix_frame && SupportsOnePass<InputColumnType, window_function> &&
+      (is_rank_like(window_function) || frame.type == FrameType::Rows))
+    return ComputationStrategy::OnePass;
+
+  if (SupportsSegmentTree<InputColumnType, window_function>)
+    return ComputationStrategy::SegmentTree;
+
+  Fail("Could not determine appropriate computation strategy for window function " +
+       window_function_to_string.left.at(window_function) + ".");
+}
+
+bool WindowFunctionEvaluator::is_output_nullable() const {
+  return !is_rank_like(_window_function_expression->window_function) &&
+         left_input_table()->column_is_nullable(_function_argument_column_id);
+}
+
+namespace {
+
+HashPartitionedData collect_chunk_into_buckets(ChunkID chunk_id, const Chunk& chunk,
+                                               std::span<const ColumnID> partition_column_ids,
+                                               std::span<const ColumnID> order_column_ids,
+                                               ColumnID function_argument_column_id) {
+  const auto chunk_size = chunk.size();
+
+  const auto collect_values_of_columns = [&chunk, chunk_size](std::span<const ColumnID> column_ids) {
+    auto values = std::vector(chunk_size, std::vector(column_ids.size(), NULL_VALUE));
+    for (auto column_id_index = 0u; column_id_index < column_ids.size(); ++column_id_index) {
```

Comment:
I think the performance argument about repeated calls to `.size()` doesn't really apply because the span (and with that, the size) is already on the function stack and the compiler can see that we are not touching it (we could make it `const` to guarantee it even further). Question now becomes whether we extract a size beforehand for performance or for consistency :thinking: (I vote performance :D)